File: CSE 3320 Labs/Lab 1/cmc3763_lab01/cmc3763_lab01.py
"""Name: Cesar Cantu-Perez
   ID# 1001463763
   Python 3.8.10
   Ubuntu VM """


"""import from dictionaries for use in the recursive function directSize"""
from pathlib import Path
import os

"""start by initializing a variable to hold the path '.' which represents current directory"""
p = Path('.')
"""the function directSize accepts two arguments: the directory and a variable for the size of the directory"""
def directSize(dir, size):
    """uses a for loop to iterate through each directory"""
    for x in dir.iterdir():
        """if x is a directory the program will call the directSize function again this time for the new directory
            and if x is not a directory the program will use the getsize function found in the os dictionary to get the
            and add the size to the current size that was passed into the directSize function  """
        if x.is_dir():
            size = directSize(x, size)
        else:
            fileSize = os.path.getsize(x)
            print(x,' is ',fileSize, 'bytes.')
            size = size + fileSize

    return size

"""program calls the directSize function to find the size of the current directory ('.')"""
directSum = directSize(p, 0)
print('The size of the directory is: ',directSum, 'bytes.')

File: CSE 3320 Labs/Lab 1/cmc3763_lab01/cmc3763_lab01.c
//Name: Cesar Cantu-Perez
//UTA ID: 1001463763
//Lang ver: GCC 8.1.0
//OS: Windows 10



/*1. It was easier and faster for me to write the code for this lab using python. Python programs are typically shorter than other programs written in a language like c. Also using directory functions were way easier for me to understand and implement in python.

  2. To simulate recursion you could use a while loop that would break once it reaches the end of what it is supposed to be doing. A downside to that approach would be that you would have to set a limit to what the result could be (i.e. a maximum number of loops).*/

#include <stdio.h>
#include <stdlib.h>
#include <dirent.h>
#include <sys/stat.h>
#include <string.h>

//function to calculate file size
int direct_size(char *directory, int sum);

int direct_size(char *directory, int sum)
{
    DIR *newDirect = opendir(directory);
	char nextDir[100];
	char fileName[100];
    struct dirent *entry;
    struct stat check;
	int comp, comp2, filesize = 1;
    //will check to see if directory has successfully been opened
    if(newDirect == NULL)
    {
        printf("Unable to access directory.");
    }
    else
    {
        //option to read all names from directory
        //printf("Directory opened!\n");
		//will read through directory adding up the size of files until it reaches a directory where it will call direct_size again
        while((entry = readdir(newDirect)) != NULL)
        {
            // get rid of the . and .. directories
			comp = strcmp(entry->d_name, ".");
            comp2 = strcmp(entry->d_name,"..");
			if(comp == 0 || comp2 == 0)
			{
				
			}
			else
			{
				//check to see if it is a directory or a file
				strcpy(fileName, directory);
				strcat(fileName, "/");
				strcat(fileName, entry->d_name);
				stat(fileName, &check);
				if(entry->d_type == DT_DIR)
				{
					//adds to the path name so direct_size can open the next directory
					printf("Directory: %s\n", entry->d_name);
					//strcpy(nextDir, directory);
					//strcat(nextDir, "/");
					//strcat(nextDir, entry->d_name);
					//printf("sum to be passed in: %d\n", sum);
					sum = direct_size(fileName, sum);
				}
				else
				{
					//adds the size to the sum of bytes already calculated
					printf("%s ", fileName);
					filesize = check.st_size;
					//printf("%d + ", sum);
					sum = sum + check.st_size;
					printf("%d bytes\n", sum);
					
				}
			}
        }
    }
	return sum;
}

int main()
{
    char direct;
	direct = '.';
	int sum = 0;
    //uses the direct as a path for the direct_size
    sum = direct_size(&direct, sum);
	printf("The size of the directory and its subdirectories is %d bytes.\n", sum);
    return 0;
}
